### PR TITLE
statistics: support historical stats dump partition table

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve?

Issue Number: close #xxx

Problem Summary:

### What is changed and how it works?

support historical stats dump partition table

### Check List

Tests 

- [x] Unit test

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: executor/analyze.go [hunk 1]
```diff
@@ -382,6 +390,12 @@ func (e *AnalyzeExec) handleResultsErrorWithConcurrency(ctx context.Context, sta
 		}
 		err = errors.New(strings.Join(errMsg, ","))
 	}
+	for tableID := range tableIDs {
+		// Dump stats to historical storage.
+		if err := recordHistoricalStats(e.ctx, tableID); err != nil {
+			logutil.BgLogger().Error("record historical stats failed", zap.Error(err))
+		}
+	}
```
Comment (xuyifangreeneyes) [inline executor/analyze.go:398]: Though we save stats concurrently by several workers, the logic of recording historical stats is sequential. Will we improve it in the future?

File: domain/historical_stats.go [hunk 2]
```diff
@@ -48,18 +49,26 @@ func (w *HistoricalStatsWorker) DumpHistoricalStats(tableID int64, statsHandle *
...
-	tblInfo := tbl.Meta()
 	dbInfo, existed := is.SchemaByTable(tblInfo)
 	if !existed {
 		return errors.Errorf("cannot get DBInfo by TableID %d", tableID)
 	}
-	if _, err := statsHandle.RecordHistoricalStatsToStorage(dbInfo.Name.O, tblInfo); err != nil {
-		generateHistoricalStatsFailedCounter.Inc()
-		return errors.Errorf("record table %s.%s's historical stats failed", dbInfo.Name.O, tblInfo.Name.O)
+	if _, err := statsHandle.RecordHistoricalStatsToStorage(dbInfo.Name.O, tblInfo, tableID, isPartition); err != nil {
+		return errors.Errorf("record table %s.%s's historical stats failed, err:%v", dbInfo.Name.O, tblInfo.Name.O, err)
```
Comment (xuyifangreeneyes) [inline domain/historical_stats.go:71]: We don't need `generateHistoricalStatsFailedCounter` here anymore?

**Review by xuyifangreeneyes — APPROVED** [review]
rest LGTM

File: domain/historical_stats.go (hunk 2 above)
Comment (Yisaer) [reply domain/historical_stats.go:72]: updated.

File: executor/analyze.go (hunk 1 above)
Comment (Yisaer) [reply executor/analyze.go:398]: dumping historical stats is worked by multiple worker in background, thus it's not sequential

File: statistics/handle/handle.go [hunk 3]
```diff
@@ -2565,17 +2565,27 @@ func (h *Handle) GetPredicateColumns(tableID int64) ([]int64, error) {
...
-func (h *Handle) RecordHistoricalStatsToStorage(dbName string, tableInfo *model.TableInfo) (uint64, error) {
+func (h *Handle) RecordHistoricalStatsToStorage(dbName string, tableInfo *model.TableInfo, physicalID int64, isPartition bool) (uint64, error) {
 	ctx := kv.WithInternalSourceType(context.Background(), kv.InternalTxnStats)
-	js, err := h.DumpStatsToJSON(dbName, tableInfo, nil, true)
+	var js *JSONTable
+	var err error
+	if isPartition {
+		js, err = h.tableStatsToJSON(dbName, tableInfo, physicalID, 0)
+	} else {
+		js, err = h.DumpStatsToJSON(dbName, tableInfo, nil, true)
```
Comment (qw4990) [inline statistics/handle/handle.go:2575]: `tableStatsToJSON` and `DumpStatsToJSON` seem very similar... Can we refactor them and merge them into one function after this PR?

**Review by qw4990 — APPROVED** [review]
LGTM

**Pull request merged**